File: android/app/src/main/java/org/gozaltech/nvdaremotecompanion/android/UpdateChecker.java
package org.gozaltech.nvdaremotecompanion.android;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;

class UpdateChecker {

    interface Callback {
        void onResult(UpdateInfo update);
        void onError(Exception e);
    }

    private static final String API_URL = "https://api.github.com/repos/gozaltech/NVDARemoteCompanion/releases/latest";

    static void checkAsync(String currentVersion, Callback callback) {
        new Thread(() -> {
            try {
                UpdateInfo result = check(currentVersion);
                callback.onResult(result);
            } catch (Exception e) {
                callback.onError(e);
            }
        }, "update-checker").start();
    }

    private static UpdateInfo check(String currentVersion) throws Exception {
        HttpURLConnection conn = (HttpURLConnection) new URL(API_URL).openConnection();
        conn.setConnectTimeout(10_000);
        conn.setReadTimeout(10_000);
        conn.setRequestProperty("Accept", "application/vnd.github+json");

        int code = conn.getResponseCode();
        if (code != 200) throw new Exception("HTTP " + code);

        StringBuilder sb = new StringBuilder();
        try (BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream()))) {
            String line;
            while ((line = br.readLine()) != null) sb.append(line);
        }

        JSONObject json = new JSONObject(sb.toString());
        String tagName = json.getString("tag_name");
        String latestVersion = tagName.startsWith("v") ? tagName.substring(1) : tagName;
        String releaseNotes = json.optString("body", "");

        if (!isNewer(latestVersion, currentVersion)) return null;

        String downloadUrl = findApkUrl(json.optJSONArray("assets"));
        if (downloadUrl == null) return null;

        return new UpdateInfo(latestVersion, downloadUrl, releaseNotes);
    }

    private static String findApkUrl(JSONArray assets) throws Exception {
        if (assets == null) return null;
        for (int i = 0; i < assets.length(); i++) {
            JSONObject asset = assets.getJSONObject(i);
            String name = asset.optString("name", "");
            if (name.endsWith(".apk")) {
                return asset.getString("browser_download_url");
            }
        }
        return null;
    }

    static boolean isNewer(String candidate, String current) {
        int[] c = parseParts(candidate);
        int[] v = parseParts(current);
        int len = Math.max(c.length, v.length);
        for (int i = 0; i < len; i++) {
            int cv = i < c.length ? c[i] : 0;
            int vv = i < v.length ? v[i] : 0;
            if (cv != vv) return cv > vv;
        }
        return false;
    }

    private static int[] parseParts(String version) {
        String[] parts = version.split("\\.");
        int[] nums = new int[parts.length];
        for (int i = 0; i < parts.length; i++) {
            try { nums[i] = Integer.parseInt(parts[i].replaceAll("[^0-9]", "")); }
            catch (NumberFormatException ignored) {}
        }
        return nums;
    }
}

File: android/app/src/main/java/org/gozaltech/nvdaremotecompanion/android/update/UpdateChecker.kt
package org.gozaltech.nvdaremotecompanion.android.update

import java.net.HttpURLConnection
import java.net.URL
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

private const val API_URL =
    "https://api.github.com/repos/gozaltech/NVDARemoteCompanion/releases/latest"
private const val TIMEOUT_MS = 10_000
private const val HTTP_OK = 200

data class UpdateInfo(
    val version: String,
    val downloadUrl: String,
    val releaseNotes: String,
)

@Serializable
private data class Release(
    @SerialName("tag_name") val tagName: String,
    val body: String = "",
    val assets: List<ReleaseAsset> = emptyList(),
)

@Serializable
private data class ReleaseAsset(
    val name: String,
    @SerialName("browser_download_url") val downloadUrl: String,
)

object UpdateChecker {

    private val json = Json { ignoreUnknownKeys = true }

    suspend fun check(currentVersion: String): Result<UpdateInfo?> =
        withContext(Dispatchers.IO) {
            runCatching {
                val release = json.decodeFromString<Release>(fetchLatestRelease())
                val latest = release.tagName.removePrefix("v")
                if (!isNewer(latest, currentVersion)) return@runCatching null
                val apk =
                    release.assets.firstOrNull { it.name.endsWith(".apk") }
                        ?: return@runCatching null
                UpdateInfo(latest, apk.downloadUrl, release.body)
            }
        }

    fun isNewer(candidate: String, current: String): Boolean {
        val candidateParts = parseVersion(candidate)
        val currentParts = parseVersion(current)
        repeat(maxOf(candidateParts.size, currentParts.size)) { index ->
            val left = candidateParts.getOrElse(index) { 0 }
            val right = currentParts.getOrElse(index) { 0 }
            if (left != right) return left > right
        }
        return false
    }

    private fun parseVersion(version: String): List<Int> =
        version.split('.').map { part -> part.filter(Char::isDigit).toIntOrNull() ?: 0 }

    private fun fetchLatestRelease(): String {
        val connection =
            (URL(API_URL).openConnection() as HttpURLConnection).apply {
                connectTimeout = TIMEOUT_MS
                readTimeout = TIMEOUT_MS
                setRequestProperty("Accept", "application/vnd.github+json")
            }
        try {
            check(connection.responseCode == HTTP_OK) { "HTTP ${connection.responseCode}" }
            return connection.inputStream.bufferedReader().use { it.readText() }
        } finally {
            connection.disconnect()
        }
    }
}
